csharp/src/ParallelSelectExt.cs

```diff
@@ -0,0 +1,374 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+using JetBrains.Annotations;
+
+namespace ArmoniK.Utils;
+
+[PublicAPI]
+public static class ParallelSelectExt
+{
+  /// <summary>
+  ///   Iterates over the input enumerable and spawn multiple parallel tasks.
+  ///   At most `parallelism` tasks will be running at any given time.
+  ///   If `parallelism` is 0, number of threads is used.
+  ///   If `parallelism` is negative, no limit is used.
+  ///   All results are collected in-order.
+  /// </summary>
+  /// <param name="enumerable">Enumerable to iterate on</param>
+  /// <param name="parallelism">Maximum number of parallel tasks to be spawned at any moment</param>
+  /// <param name="cancellationToken">Token to cancel the enumeration</param>
+  /// <typeparam name="T">Return type of the tasks</typeparam>
+  /// <returns>Asynchronous results of tasks</returns>
+  [PublicAPI]
+  public static async IAsyncEnumerable<T> ParallelWait<T>(this IEnumerable<Task<T>>                  enumerable,
+                                                          int                                        parallelism,
+                                                          [EnumeratorCancellation] CancellationToken cancellationToken = default)
+  {
+    // If parallelism is 0, use number of threads
+    if (parallelism == 0)
+    {
+      parallelism = Environment.ProcessorCount;
+    }
+
+    // If parallelism is negative, launch as many tasks as possible
+    if (parallelism < 0)
+    {
+      parallelism = int.MaxValue;
+    }
+
+    // cancellation task for early exit
+    var cancelled = cancellationToken.AsTask<T>();
+    // Queue of tasks
+    var queue = new Queue<Task<T>>();
+    // Semaphore to limit the parallelism
+    using var sem = new SemaphoreSlim(parallelism - 1, parallelism);
+
+    // Iterate over the enumerable
+    foreach (var x in enumerable)
+    {
+      // Prepare the new semaphore acquisition
+      var semAcquire = sem.WaitAsync(cancellationToken);
+
+      // Async closure that waits for the task x and releases the semaphore
+      async Task<T> TaskLambda()
+      {
+        var res = await x.ConfigureAwait(false);
+        sem.Release();
+        return res;
+      }
+
+      // We can enqueue the task
+      queue.Enqueue(Task.Run(TaskLambda, cancellationToken));
+
+      // Dequeue tasks as long as the semaphore is not acquired yet
+      while (true)
+      {
+        var front = queue.Count != 0
+                      ? queue.Peek()
+                      : null;
+        // Select the first task to be ready
+        var which = await Task.WhenAny(cancelled,
+                                       front ?? cancelled,
+                                       semAcquire)
+                              .ConfigureAwait(false);
+
+        // Cancellation has been requested
+        if (which.IsCanceled)
```

Comment:
Should not be required. If which is canceled, it should have thrown in the await.

Reply:
However, you may want to cancel all the other running tesks.

Reply:
That's not how `Task.WhenAny` works. `WhenAny` returns the first task that is completed (either ran to completion, cancelled, or failed). You are expected to await it afterward. I might add an extension to do just that, like a `cancellationToken.ThrowIfCancelled()`, but for tasks.